binderhub/build.py

```diff
@@ -144,8 +150,59 @@ def progress(self, kind, obj):
         """Put the current action item into the queue for execution."""
         self.main_loop.add_callback(self.q.put, {'kind': kind, 'payload': obj})
 
+    def get_affinity(self):
+        """Determine the affinity term for the build pod.
+
+        There are a total of two affinity strategies, which one is used depends
+        on how the BinderHub is configured.
+
+        In the default setup the affinity of each build pod is an "anti-affinity"
+        which causes pods to prefer to schedule on separate nodes.
+
+        In a setup with docker-in-docker pods for a particular repository
+        prefer to schedule on the same node in order to reuse the build cache
+        of previous builds.
+        """
+        dind_pods = self.api.list_namespaced_pod(
+            self.namespace,
+            label_selector="component=dind,app=binder",
+        )
+        if self.sticky_builds and dind_pods:
+            nodes = dict((item.spec.node_name, item.spec)
+                         for item in dind_pods.items)
```

Comment:
First time i see dictionary comprehension like this, and perhaps a different naming of the variable?

```python
dind_pods_node_affinity = {
    item.spec.node_name: item.spec.affinity.node_affinity
    for item in dind_pods.items()
}
```

But hmmm... I don't understand how we can reuse the dind pods affinity term? Oh!!! Is it because it is a daemonset and has an automatic hard affinity for the specific hostname? Clever! That should be clarified though!

Hmmmm but hmmm... there could be additional hard affinities set on these pods. Also, required and preferred has a different schema doesn't they? Yes, see: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.15/#nodeaffinity-v1-core

Both will share a NodeSelectorTerm object type though. I suggest we search for the NodeSelectorTerm object within `required_during_scheduling_ignored_during_execution.nodeSelectorTerms` that references a kubernetes.io/hostname label with some value. A helper function could be in place for this. Then, we make two entries for the top two ranked nodes where the highest ranked node gets a weight of 100 and the other a weight of 50.

Reply:
I switched to explicitly creating the affinity spec.

After your comments and digging into things I realised trying to be clever here was not worth it. Constructing the affinity spec is tedious to type once (now) but easier to understand in the future.